### PR TITLE
Test xrootd with multiple authentication methods, including VOMS

### PR DESCRIPTION
**Pull request by matyasselmeci**
VMU tests forthcoming.

### CODE REVIEW

**Comment by matyasselmeci** [comment]
Tests: https://osg-sw-submit.chtc.wisc.edu/tests/20211104-0928/results.html
Those are all the failures I expected.

File: osgtest/tests/test_150_xrootd.py [hunk 1]
```diff
@@ -28,20 +32,21 @@
...
 # All paths are relative to the rootdir defined above
-AUTHFILE_TEXT = """\
+AUTHFILE_TEXT = f"""\
 # A user has full privileges to a directory named after them (e.g. matyas has /matyas/, vdttest has /vdttest/)
 u =      /@=/ a
 
-# The xrootd user has full privileges to the top-level directory
-u xrootd / a
+# Our test VO has full privileges to /osgtestvo
+g /{voms.VONAME} /osgtestvo/ a
```
Comment (brianhlin) [inline osgtest/tests/test_150_xrootd.py:46]:
```suggestion
g /{voms.VONAME} /{voms.VONAME}/ a
```

Maybe?

File: osgtest/tests/test_150_xrootd.py [hunk 2]
```diff
@@ -72,34 +77,36 @@ def setUp(self):
...
         core.config['xrootd.config'] = '/etc/xrootd/config.d/10-osg-test.cfg'
         core.config['xrootd.logging-config'] = '/etc/xrootd/config.d/99-logging.cfg'
         core.config['xrootd.service-defaults'] = '/etc/sysconfig/xrootd'
         core.config['xrootd.multiuser'] = False
-        core.config['xrootd.ztn'] = False
         core.state['xrootd.backups-exist'] = False
         core.state['xrootd.had-failures'] = False
         core.config['xrootd.public_subdir'] = "public"
         core.config['xrootd.user_subdir'] = core.options.username
+        core.config['xrootd.vo_subdir'] = "osgtestvo"
```
Comment (brianhlin) [inline osgtest/tests/test_150_xrootd.py:92]:
```suggestion
        core.config['xrootd.vo_subdir'] = voms.VONAME
```

File: osgtest/tests/test_158_xrootd_tpc.py [hunk 3]
```diff
@@ -92,9 +92,9 @@ def test_01_configure_xrootd(self):
                     XROOTD_CFG_TEXT,
                     owner='xrootd', backup=True, chown=(user.pw_uid, user.pw_gid),
                     chmod=0o644)
-        files.write('/etc/xrootd/config.d/40-osg-standalone.cfg', XROOTD_STANDALONE_TXT,
-                    owner='xrootd', backup=True, chown=(user.pw_uid, user.pw_gid),
-                    chmod=0o644)
+        # files.write('/etc/xrootd/config.d/40-osg-standalone.cfg', XROOTD_STANDALONE_TXT,
+        #             owner='xrootd', backup=True, chown=(user.pw_uid, user.pw_gid),
+        #             chmod=0o644)
```
Comment (brianhlin) [inline osgtest/tests/test_158_xrootd_tpc.py:97]: Why not just remove this entirely?

File: osgtest/tests/test_838_xrootd_tpc.py [hunk 4]
```diff
@@ -24,7 +24,7 @@ def test_02_stop_xrootd_tpc(self):
             files.restore(core.config['xrootd.tpc.config-1'], "xrootd")
             files.restore(core.config['xrootd.tpc.config-2'], "xrootd")
             files.restore(core.config['xrootd.tpc.basic-config'], "xrootd")
-            files.restore('/etc/xrootd/config.d/40-osg-standalone.cfg', "xrootd")
+            # files.restore('/etc/xrootd/config.d/40-osg-standalone.cfg', "xrootd")
```
Comment (brianhlin) [inline osgtest/tests/test_838_xrootd_tpc.py:27]: Why not just remove this entirely?

**Review by brianhlin — COMMENTED** [review]
Had time to review a bit of this

File: osgtest/tests/test_158_xrootd_tpc.py (hunk 3 above)
Comment (matyasselmeci) [reply osgtest/tests/test_158_xrootd_tpc.py:97]: Yep

File: osgtest/tests/test_401_scitokens.py [hunk 5]
```diff
@@ -21,13 +21,13 @@ def test_01_request_condor_write_scitoken(self):
...
     def test_02_request_xrootd_scitokens(self):
-        self.skip_ok_unless("SCITOKENS" in core.config.get('xrootd.security', []),
+        self.skip_ok_unless("SCITOKENS" in core.config.get('xrootd.security', set()),
                             "Not using SciTokens for XRootD")
         try:
             credentials.request_scitoken(
                 "xrootd",
                 subject=core.options.username,
-                scope="write:/",
+                scope=f"read:/ write:/{core.config['xrootd.user_subdir']}",
```
Comment (brianhlin) [inline osgtest/tests/test_401_scitokens.py:30]: Should we also restrict read on this token or is restricted access of VO area restricted by the authfile?

File: osgtest/tests/test_450_xrootd.py [hunk 6]
```diff
@@ -11,188 +11,265 @@
...
+        self.skip_unless_security("SCITOKENS")
+        xrootd_url = xroot_url(TestXrootd.user_copied_file_scitoken)
+        command = ('xrdcp', '--force', '--debug', '2', TestXrootd.__data_path, xrootd_url)
+        with core.no_x509(core.options.username):
+            with core.environ_context({"BEARER_TOKEN": core.state['token.xrootd_contents'], "BEARER_TOKEN_FILE": None}):
+                message = "xrdcp upload to user dir with scitoken in BEARER_TOKEN"
+                expected_exit = 0
+                if core.PackageVersion("xrootd-libs") <= "5.3.2":
+                    message += " (expected failure)"
+                    expected_exit = ERR_AUTH_FAIL
```
Comment (brianhlin) [inline osgtest/tests/test_450_xrootd.py:132]:
```suggestion
                if core.PackageVersion("xrootd-libs") <= "5.3.2":
                    message += " (expected failure)"
                    expected_exit = ERR_AUTH_FAIL
```

A comment for why this is the case would be helpful

File: osgtest/tests/test_465_xrootd_tpc.py [hunk 7]
```diff
@@ -84,78 +103,65 @@ def test_01_create_macaroons(self):
...
+        with core.no_x509(core.options.username), core.no_bearer_token(core.options.username):
             core.system(command, user=True)
-            time.sleep(1)
-            self.assertTrue(os.path.exists(dest_path), "Copied file missing")
-            self.assertTrue(files.checksum_files_match(TestXrootdTPC.source_path, dest_path),
-                            "Copied file contents do not match original")
-        except AssertionError:
-            core.state['xrootd.tpc.had-failures'] = True
-            raise
+        time.sleep(1)
```
Comment (brianhlin) [inline osgtest/tests/test_465_xrootd_tpc.py:115]:
```suggestion
        time.sleep(1)
```

A comment explaining why would be helpful

**Review by brianhlin — APPROVED** [review]
LGTM. Some minor questions/comments

File: osgtest/tests/test_401_scitokens.py (hunk 5 above)
Comment (matyasselmeci) [reply osgtest/tests/test_401_scitokens.py:30]: Let's restrict read access as well. I think the authfile is ignored for tokens.

File: osgtest/tests/test_450_xrootd.py (hunk 6 above)
Comment (matyasselmeci) [reply osgtest/tests/test_450_xrootd.py:132]: I'll file a bug report and add a comment pointing to the issue.

**Comment by matyasselmeci** [comment]
Still not having luck with the xrootd voms tests using voms-proxy-direct: https://osg-sw-submit.chtc.wisc.edu/tests/20211109-2237/packages.html

Any ideas?

**Comment by matyasselmeci** [comment]
Actually I have one: the vomses and the lsc file don't get set up unless the server packages are installed. I'll try changing that.

**Comment by matyasselmeci** [comment]
@brianhlin's latest tests look good: https://osg-sw-submit.chtc.wisc.edu/tests/20211110-1459/packages.html

**Pull request merged**